sycl/doc/Assert.md

```diff
@@ -0,0 +1,453 @@
+text of the failing expression, the values of the standard macros `__FILE__` and
+`__LINE__`, and the value of the standard variable `__func__`. If the failing
+assert comes from an `nd_range` `parallel_for` it will also include the global
+ID and the local ID of the failing work item.
+
+Implementation of this function is supplied by Native Device Compiler for
+safe approach or by DPCPP Compiler for fallback one.
+
+NB: Due to lack of support of online linking in Level-Zero, the application is
+linked against fallback implementation of `__devicelib_assert_fail`. Hence,
+Native Device Compilers should prefer their implementation instead of the one
+provided in incoming SPIR-V/LLVM IR binary.
+
+Limitations for user after catching the "assert" asynchronous exception:
+ - When using GPU device and the kernel hangs/crashes the subsequent enqueues
+   will fail;
+
+When using CPU devices the user can proceed with enqueues to the same
+device/queue/context.
+DPCPP Runtime remains in valid state after "assert" exception been thrown.
+
+
+### Current violation
+
+While throwing an asynchronous exception is quite an extensible way, for the
+time being DPCPP Runtime merely calls `abort()`.
+
+
+## Safe approach
+
+This is the preferred approach and implementations should use it when possible.
+It guarantees assertion failure notification delivery to the host regardless of
+kernel behavior which hit the assertion.
+
+The Native Device Compiler is responsible for providing implementation of
+`__devicelib_assert_fail` which completely hides details of communication
+between the device code and the Low-Level Runtime from the SYCL device compiler
+and runtime. The Low-Level Runtime is responsible for:
+ - detecting if assert failure took place;
+ - flushing assert message to `stderr` on host.
+
+When detected, Low-level Runtime reports assert failure to DPCPP Runtime
+via events objects. To achieve this, information about assert failure should be
+propagated from device-side to SYCL Runtime. This should be performed via calls
+to `piEventGetInfo`. This Plugin Interface call "lowers" to `clGetEventInfo` for
+OpenCL backend and `zeEventQueryStatus` for Level-Zero backend.
+
+Refer to [OpenCL](extensions/Assert/opencl.md) and [Level-Zero](extensions/Assert/level-zero.md)
+extensions.
+
+The following sequence of events describes how user code gets notified:
+ - Device side:
+   1. Assert fails in device-code in kernel
+      // It's not defined if GPU thread stops execution
+      // Other GPU threads are left untouched
+   2. Specialized version of `__devicelib_assert_fail` is called
+   3. Device immediately signals to host (Low-Level Runtime)
+ - Host side:
+   1. The assert failure gets detected by Low-Level Runtime
+   2. Low-Level Runtime sets event status
+   3. Upon call to `sycl::queue::wait_and_throw()` or
+      `sycl::event::wait_and_throw()` DPCPP Runtime checks event status and
+      throws "assert" exception
```

Comment:
Thanks.
In the light of recent offline discussion regarding throwing exceptions vs immediate abort - host side should be changed to the following, I believe:
1. ...
2. Low level runtime prints the assert message and calls abort()

+ @GarveyJoe, @kbsmith-intel

Reply:
@kbobrovs  - Yes, I agree.

Reply:
It was previously stated in the document:
> While throwing an asynchronous exception is quite an extensible way, for the time being DPCPP Runtime merely calls `abort()`.

Doesn't this fit?

Reply:
That statement makes it sound like the intended interface is that we will throw an asynchronous exception but the current implementation has a bug (i.e. it calls `abort` instead).  If the intention is that the interface will change to call `abort`, we should change the extension specification to say this (and remove all the stuff about throwing an asynchronous exception).

I have no concerns about calling `abort` instead of throwing the exception.  This is more consistent with the host version of `assert`.

Reply:
Should the events have some erroneous status, then?

Reply:
@s-kanaev, I think the asynchronicity design needs quite a bit of elaboration anyway, and can be done later.  For now let's follow the approach that reviewers recall as the one chosen during the offline discussion - abort() w/o any asynchronous stuff for now. I think this would simplify the process a lot and will allow to add complexity incrementally. This would need to be done then:

 > If the intention is that the interface will change to call abort, we should change the extension specification to say this (and remove all the stuff about throwing an asynchronous exception).

Reply:
And, BTW, thanks for patiently addressing all the comments :)

Reply:
> This would need to be done then:
>> If the intention is that the interface will change to call abort, we should change the extension specification to say this (and remove all the stuff about throwing an asynchronous exception).

Already done.

Reply:
Given this direction, can we delete the cl_intel_assert_return_code and ze_intel_assert_return_code extensions completely?  They are only needed if we want the SYCL runtime instead of the low level runtime to call abort.

Reply:
Sounds reasonable.
Removed.